homeassistant/components/goodwe/sensor.py

```diff
@@ -0,0 +1,252 @@
+        self._attr_unique_id = f"{DOMAIN}-{inverter.serial_number}"
+        self._attr_device_info = device_info
+        self._attr_icon = "mdi:solar-power"
+        self._attr_name = "PV Inverter"
+        self._attr_native_unit_of_measurement = POWER_WATT
+        self._attr_extra_state_attributes = {"serial_number": inverter.serial_number}
+        self._inverter: Inverter = inverter
+        self._previous_value = None
+
+    async def async_set_work_mode(self, work_mode: int) -> None:
+        """Set the inverter work mode."""
+        await self._inverter.set_work_mode(work_mode)
+
+    async def async_set_ongrid_battery_dod(self, ongrid_battery_dod: int) -> None:
+        """Set the on-grid battery dod."""
+        await self._inverter.set_ongrid_battery_dod(ongrid_battery_dod)
+
+    async def async_set_grid_export_limit(self, grid_export_limit: int) -> None:
+        """Set the grid export limit."""
+        await self._inverter.set_grid_export_limit(grid_export_limit)
+
+    @property
+    def native_value(self):
+        """Return the value reported by the sensor."""
+        new_value = self.coordinator.data.get(
+            self._MAIN_ENTITY_SENSOR, self._previous_value
+        )
+        self._previous_value = new_value
+        return new_value
+
+    @property
+    def state_attributes(self):
+        """Return the inverter state attributes."""
+        return self.coordinator.data
+
+
+class InverterSensor(CoordinatorEntity, SensorEntity):
+    """Entity representing individual inverter sensor."""
+
+    def __init__(
+        self,
+        coordinator: DataUpdateCoordinator,
+        device_info: DeviceInfo,
+        inverter: Inverter,
+        sensor: Sensor,
+    ) -> None:
+        """Initialize an inverter sensor."""
+        super().__init__(coordinator)
+        self.entity_id = f".{DOMAIN}_{sensor.id_}"
+        self._attr_unique_id = f"{DOMAIN}-{sensor.id_}-{inverter.serial_number}"
+        self._attr_name = sensor.name.strip()
+        self._attr_device_info = device_info
+        self._attr_entity_category = (
+            ENTITY_CATEGORY_DIAGNOSTIC if sensor.id_ not in _MAIN_SENSORS else None
+        )
+        if sensor.unit in _DESCRIPTIONS:
+            self.entity_description = _DESCRIPTIONS[sensor.unit]
+        else:
+            self._attr_native_unit_of_measurement = sensor.unit
+        self._attr_icon = _ICONS.get(sensor.kind)
+        # Set the inverter SoC as main device battery sensor
+        if sensor.id_ == BATTERY_SOC:
+            self._attr_device_class = DEVICE_CLASS_BATTERY
+        self._sensor = sensor
+        self._previous_value = None
+
+    @property
+    def native_value(self):
+        """Return the value reported by the sensor."""
+        new_value = self.coordinator.data.get(self._sensor.id_, self._previous_value)
+        # Total increasing sensor should never be set to 0
+        if (
+            self.state_class == STATE_CLASS_TOTAL_INCREASING
```

Comment:
I'd move this case to the entity description in an optional lambda function. The lambda can accept the new value and the previous value. Example:

https://github.com/home-assistant/core/blob/c02aae58fb49963bce070e512e24a765434f7e5c/homeassistant/components/airly/sensor.py#L115

Reply:
Done, lambda was added to entity description and the special version added to kWh entity.